File: godot/scripts/orbital/Propagator.cs
// Compliance Note
// This project is based entirely on publicly available academic information and general knowledge of orbital mechanics. It contains no restricted, proprietary, or export-controlled information of any kind. This is a personal learning project only.

namespace DeltaWrecker.Orbital;

/// <summary>
/// Simplified 2D Keplerian circular-orbit propagator.
/// Mirrors engine/src/orbital/propagator.rs.
///
/// Given an initial <see cref="OrbitState"/> and a time delta Δt (seconds),
/// computes the new position and velocity for a stable near-circular orbit.
/// No burns or maneuvers are simulated at this stage.
/// </summary>
public static class Propagator
{
    /// <summary>
    /// Propagates a circular orbit forward by <paramref name="dtSeconds"/> seconds.
    ///
    /// Algorithm (same as Rust engine):
    ///   1. Compute angular velocity ω = v / r
    ///   2. Compute rotation angle θ = ω · Δt
    ///   3. Rotate the position vector by θ in the XY plane
    ///   4. Derive new velocity perpendicular to the new position
    /// </summary>
    /// <param name="state">Current orbit state (position + velocity).</param>
    /// <param name="dtSeconds">Time step in seconds.</param>
    /// <returns>New orbit state after propagating by <paramref name="dtSeconds"/>.</returns>
    public static OrbitState Propagate(OrbitState state, double dtSeconds)
    {
        // TODO-Coder: Implement circular orbit propagation to match propagator.rs.
        // Use the same algorithm:
        //   double r = state.DistanceKm();
        //   double v = state.SpeedKms();
        //   double omega = v / r;
        //   double theta = omega * dtSeconds;
        //   double cosTheta = Math.Cos(theta);
        //   double sinTheta = Math.Sin(theta);
        //   double x  =  state.X * cosTheta - state.Y * sinTheta;
        //   double y  =  state.X * sinTheta + state.Y * cosTheta;
        //   double z  =  state.Z;
        //   double vx = -v * y / r;
        //   double vy =  v * x / r;
        //   double vz =  state.Vz;
        //   return new OrbitState(x, y, z, vx, vy, vz);
        throw new NotImplementedException("Propagate not yet implemented.");
    }
}

File: engine/src/orbital/propagator.rs
// **Compliance Note**
// This project is based entirely on publicly available academic information and general knowledge of orbital mechanics. It contains no restricted, proprietary, or export-controlled information of any kind. This is a personal learning project only.

//! Simplified 2D Keplerian circular orbit propagator.
//!
//! Propagates a given initial state and Δt to compute a new position and velocity
//! for a stable near-circular orbit. No burns or maneuvers are simulated.

use super::OrbitState;

/// Propagates a circular orbit by a given time delta dt (in seconds).
/// 
/// Note: Pure time-based propagation. Given the initial state and Δt, 
/// computes new position + velocity for a stable near-circular orbit.
pub fn propagate(state: &OrbitState, dt_s: f64) -> OrbitState {
    // - Matches analytical circular orbit math within 1e-6 error
    // - Frame-rate independent
    
    let r = state.distance_km(); // distance from center of Earth to the orbiting body
    let v: f64 = state.speed_kms(); // speed of the orbiting body in km/s
    let omega = v / r; // angular velocity of the orbiting body in rad/s
    let theta = omega * dt_s; // number of radians the spacecraft moves along its circular orbit over dt_s seconds
    let cos_theta = theta.cos(); 
    let sin_theta = theta.sin();

    let x = state.position().x * cos_theta - state.position().y * sin_theta;
    let y = state.position().x * sin_theta + state.position().y * cos_theta;
    let z = state.position().z;
    
    let vx = -v * y / r;
    let vy = v * x / r;
    let vz = state.velocity().z;

    OrbitState::new(x, y, z, vx, vy, vz)
}

// **Compliance Note**
// This project is based entirely on publicly available academic information and general knowledge of orbital mechanics. It contains no restricted, proprietary, or export-controlled information of any kind. This is a personal learning project only.
